docs/compiler/architecture.md

```diff
@@ -0,0 +1,183 @@
+
+Here's a rough diagram of the compiler's architecture:
+
+```mermaid
+flowchart
+  source[Source Program]
+  ast_and_definitions[AST & Definitions]
+  monomorphized_ast[Monomorphized AST]
+  ownership_ast[Monomorphized AST with Clones & Drops]
+  optimized_ssa[Optimized SSA]
+  brillig[Brillig IR]
+
+  source --Lexing--> Tokens --Parsing--> AST --Definition Collection--> ast_and_definitions --Elaboration--> HIR
+
+  HIR --Monomorphization--> monomorphized_ast --Ownership Analysis--> ownership_ast --SSA-gen--> SSA
+
+  SSA --SSA Optimizations--> optimized_ssa
+
+  optimized_ssa --Brillig-gen--> brillig
+  optimized_ssa --ACIR-gen--> ACIR
+```
+
+Below is a summary of each compiler pass and which Rust source files may be most useful in
+understanding the pass.
+
+## Lexing
+
+Lexing is a straightforward pass which converts a single source file or stream of text into a vector
+of tokens. A token can be a single word, number, or operator. This pass can be found in
+`compiler/noirc_frontend/src/lexer/lexer.rs`.
+
+## Parsing
+
+Parsing converts the vector of tokens from the lexer into an abstract syntax tree (the
+`ParsedModule`). Noir uses a hand-written recursive descent parser for better error recovery and
+error messages. The parser can be found in `compiler/noirc_frontend/src/parser/parser.rs`.
+
+Note that the resulting AST is generally expected to maintain most information about how the source
+program is structured so that the formatter can use this to format files. The main exception to this
+is comments which must be recovered with additional help from the lexer.
+
+## Definition Collection
+
+Definition Collection is a short step where any top-level definitions are registered for future use
+during elaboration. This pass can be found in `compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs`
+and `compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs`.
+
+## The Elaborator
+
+The elaborator is the largest pass in the frontend because it performs multiple tasks simultaneously.
+It traverses the AST performing name resolution and type checking of each node in tandem (e.g. we
+perform both for one statement before moving on to the next). This is to support the third task the
+elaborator performs which is executing comptime code. Comptime code must be resolved and type
+checked before it is executed, but may also output more code which needs to be resolved and type
+checked, so it is included as part of elaboration. The elaborator outputs `Hir` nodes which are
+stored within the `NodeInterner`. The elaborator can be found in `compiler/noirc_frontend/src/elaborator/mod.rs`.
+
+Comptime code is executed by the comptime interpreter which notably can call back into the
+elaborator to resolve more code to support comptime noir functions like `Expr::resolve`.
+The comptime interpreter can be found in `compiler/noirc_frontend/src/hir/comptime/interpreter.rs`.
```

Comment:
Did you deliberately avoid using links? I'm thinking with links it could be easier to navigate, and potentially also to check that the links are valid, either automatically, or just by clicking. An automation _could_ try to guess that if it looks like a link it should work, but then why not go the full monty and help the user not have to go to the file finder and type in the path.

Reply:
No, I just didn't think of it. I'll update them